apollo-federation/src/operation/mod.rs

```diff
@@ -789,7 +799,7 @@ impl Selection {
     pub(crate) fn from_element(
         element: OpPathElement,
         sub_selections: Option<SelectionSet>,
-        unnecessary_directives: Option<&DirectiveList>,
+        unnecessary_directives: Option<&HashSet<Node<apollo_compiler::ast::Directive>>>,
```

Comment:
why `HashSet` and not just `DirectiveList`?

Reply:
Yes, this avoids duplicates, but more importantly, `DirectiveList`s don't provide a way to be mutated. You must allocate a new list if you want to "mutate" it.